### ISSUE TITLE
An empty `allow-interfaces=` makes the system not reachable

### ISSUE THREAD

**Issue by votdev** [lathiat/avahi#536]
> If set to the empty list all local interfaces except loopback and point-to-point will be used. 

Even if the documentation tells that it should work, it simply does not. I need to uncomment the `allow-interfaces=` line in `/etc/avahi/avahi-daemon.conf` and restart the service with `systemctl restart avahi-daemon.service` to make the system reachable again.

The problem occurs on Debian 12 with the following package versions.

```
# dpkg -l | grep avahi
ii  avahi-daemon                    0.8-10                               amd64        Avahi mDNS/DNS-SD daemon
ii  libavahi-client3:amd64          0.8-10                               amd64        Avahi client library
ii  libavahi-common-data:amd64      0.8-10                               amd64        Avahi common data files
ii  libavahi-common3:amd64          0.8-10                               amd64        Avahi common library
ii  libavahi-core7:amd64            0.8-10                               amd64        Avahi's embeddable mDNS/DNS-SD library
```

**Comment by votdev** [comment]
Sorry, i missed the important word `local` in the docs. So this issue can be closed.

**Comment by votdev** [comment]
I'm sorry, i missed the `all local interfaces` in the docs.

**Comment by votdev** [comment]
After some confusion about what `local` means I have to reopen the issue.

**Comment by evverx** [comment]
Looks like it's a documentation bug. It should probably say that it should be unset to pick up the default interfaces. It's unset in the default config coming with avahi.

**Comment by tomeq82** [comment]
Exactly, this is documentation/functionality bug - empty string "allow-interfaces" makes NO interfaces catched/enumerated thus virtually avahi not working at all. No matter which distro, Avahi 0.8

**Comment by damiank89** [comment]
In my opinion, this isn't a documentation bug. It's expected for an empty value in both `allow-interfaces` and `deny-interfaces` to set the corresponding `server_config` value to `NULL`, but this isn't the case because `avahi_split_csv()` does not return `NULL` for an empty string input. The result is a list of one item (an empty string), instead of an empty list (`NULL`).

This is what causes the bug. `allow-interfaces` is essentially set to one interface, with an empty name.

Not a documentation bug for sure, and can be solved in three ways:
1. Error out when empty values are provided (#833)
2. Warn and skip the line when empty value is provided
3. Change `avahi_split_csv()` to return `NULL` when given an empty string

I think this should be solved with an error, but a warning is also acceptable. Changing `avahi_split_csv()` probably requires more work and testing, since it's a core function.

@evverx, I'm open to more suggestions if you have any.

**Comment by evverx** [comment]
> can be solved in three ways

What do you think about https://github.com/avahi/avahi/issues/439? Would it change the way the empty values should be handled? What do you think about the convention like the one mentioned in https://www.man7.org/linux/man-pages/man7/systemd.syntax.7.html
> Various settings are allowed to be specified more than once, in
       which case the interpretation depends on the setting. Often,
       multiple settings form a list, and setting to an empty value
       "resets", which means that previous assignments are ignored.

**Comment by votdev** [comment]
In my opinion, the default behavior should simply be applied when the value is empty.

**Comment by evverx** [comment]
Agreed. I'm not sure how safe it is to assume that the undocumented behavior isn't relied on. For example the first config I found was https://github.com/truenas/middleware/blob/c082ebb17d51fd7b8c735963e444bd031a9b494f/src/middlewared/middlewared/etc_files/local/avahi/avahi-daemon.conf.mako#L43-L44.
```sh
deny-interfaces=${", ".join(deny_interfaces)}
allow-interfaces=${", ".join([x["name"] for x in allow_interfaces])}
```
can expand to the empty string there as far as I understand and it's hard to tell what is expected there. That being said the undocumented behavior shouldn't be relied on.

**Comment by evverx** [comment]
Either way my understanding is that @damiank89 is interested in addressing this so it would be great if all those things including potential future config extensions were taken into account one way or another.

**Comment by damiank89** [comment]
Yeah, I'm probably going to try another approach for this. Will also look into #439. You can assign both issues to me if you'd like.

**Comment by evverx** [comment]
I don't think it would be necessary. Either way before opening PRs like https://github.com/avahi/avahi/pull/833 it should all be discussed in issues.

**Comment by damiank89** [comment]
@evverx I've pushed a simpler fix for this issue to my branch. I'm also working on a more generic solution for empty values, but that might take a while as it requires more manual validations.

**Comment by evverx** [comment]
I'm not sure why the avahi config is suddenly so popular these days but given that https://github.com/avahi/avahi/pull/841 has just been opened I think it's a great opportunity to discuss those things and how they fit together.
